### PR TITLE
Site Editor: Offset insertion point in iframe

### LINKED ISSUE #28457: Site Editor: Columns block "add column" quick inserter showing up at the wrong location
## Description
Quick inserter shows up at the wrong location when trying to add a `Column` between two `Column` in the `Columns` block. This issue only happens in the post editor.

## Step-by-step reproduction instructions
1. Insert `Paragraph` block, fill it up with some random text
2. Insert `Columns` block, choose 50 / 50
3. Move your cursor between the two `Column` block
4. Plus button shows above the block

## Expected behaviour
Plus button should pop up between the two column block
## Actual behaviour
Plus button pops up above the Columns block

## Screenshots or screen recording (optional)

https://user-images.githubusercontent.com/2256104/105689815-853b7280-5efb-11eb-9908-c27e518789f3.mov

## WordPress information
- WordPress version: `master` branch
- Gutenberg version: `master` branch

## Device information
- Device: Macbook Pro
- Operating system: OSX
- Browser: Chrome

### PR DESCRIPTION
**Pull request by david-szabo97**
## Description
Fixes https://github.com/WordPress/gutenberg/issues/28457 https://github.com/WordPress/gutenberg/issues/28885
We moved the editor content into an iframe. It already handles offsetting the popover if a ref is passed. But insertion point calculates the position on its own which doesn't take the iframe offset into account. This change offsets the insertion point in the iframe so it's placed at the right position.

## How has this been tested?
1. Insert `Paragraph` block, fill it up with some random text
2. Insert `Columns` block, choose 50 / 50
3. Move your cursor between the two `Column` block
4. Plus button at the right position

## Screenshots

**Before**

https://user-images.githubusercontent.com/2256104/105689815-853b7280-5efb-11eb-9908-c27e518789f3.mov

**After**

https://user-images.githubusercontent.com/2256104/106460481-b552ba80-6493-11eb-9746-96934873614f.mov

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/block-list/insertion-point.js [hunk 1]
```diff
@@ -23,6 +23,24 @@ import { isRTL } from '@wordpress/i18n';
 import Inserter from '../inserter';
 import { getBlockDOMNode } from '../../utils/dom';
 
+function offsetIframe( rect, ownerDocument ) {
```
Comment (david-szabo97) [inline packages/block-editor/src/components/block-list/insertion-point.js:26]: This functions comes from https://github.com/WordPress/gutenberg/blob/e54a9ec2f97e28db3293628e921e40e81cfe5f76/packages/components/src/popover/index.js#L45

Is there a good package to host this function so we can use it in both `components` and `block-editor` package?

Comment (mattwiebe) [reply packages/block-editor/src/components/block-list/insertion-point.js:26]: I'm not sure if there's an clear candidate but maybe the `@wordpress/dom` package could be a fit.

**Review by mattwiebe — APPROVED** [review]
I tested this in the post editor (no iframe) and site editor (iframe) and it works as advertised. We should be able to clean some conditionals up once we use the iframe everywhere, but this does the job for now.

**Comment by jeyip** [comment]
## Testing
### Behavior
- [x] Offsets the insertion point in columns correctly

### Browsers
- [x] Firefox
- [x] Edge
- [x] Chrome
- [x] Safari
- [ ] ~IE11~

It looks like the site editor isn't loading again in IE11. It isn't caused by this PR because I can confirm the same issue on the master branch.

Screen Shot 2021-02-05 at 11 45 43 AM

**Comment by jeyip** [comment]
> We should be able to clean some conditionals up once we use the iframe everywhere, but this does the job for now.

I agree -- I think it's fine to move forward with this for now, but could we create an issue and/or leave a comment explaining the `offsetIframe` function and that we plan return to it when the iframe is used everywhere?

**Comment by jeyip** [comment]
> We moved the editor content into an iframe. It already handles offsetting the popover if a ref is passed. But insertion point calculates the position on its own which doesn't take the iframe offset into account.

`NB:` This seems strange. I wonder why the insertion point needs to calculate its own position if the editor is already handling positioning of popovers. It feels like this kind of offsetting logic should be handled by a single source of truth.

File: packages/block-editor/src/components/block-list/insertion-point.js (hunk 1 above)
Comment (jeyip) [reply packages/block-editor/src/components/block-list/insertion-point.js:26]: I took a brief look at Gutenberg packages and agree that `@wordpress/dom` might be the right place

**Review by jeyip — COMMENTED** [review]
+1 from me when we resolve the conversation about DRYing out offsetIframe from the `components` and `block-editor` packages

**Comment by david-szabo97** [comment]
* Moved `offsetIframe` into `@wordpress/dom`
* Added tests. I wonder what you think about these tests. JSDom doesn't have a layout engine so I had to mock `getBoundingClientRect` and `DOMRect`.

File: packages/dom/src/dom.js [hunk 2]
```diff
@@ -1002,3 +1002,29 @@ export function removeInvalidHTML( HTML, schema, inline ) {
...
+/**
+ * Offsets the given rect by the position of the iframe that contains the element.
+ * If the owner document is not in an iframe then it returns with the original rect.
+ *
+ * @param {DOMRect} rect bounds of the element
+ * @param {Document} ownerDocument document of the element
+ *
+ * @return {DOMRect} offsetted bounds
+ */
+export function offsetIframe( rect, ownerDocument ) {
```
Comment (youknowriad) [inline packages/dom/src/dom.js:1015]: now that it's a public API, maybe we should find a better name? @ellatrix any thoughts on this?

File: packages/dom/src/test/dom.js [hunk 3]
```diff
@@ -200,6 +201,70 @@ describe( 'DOM', () => {
...
+			const iframeLeft = 25;
+			const iframeTop = 50;
+			const childLeft = 10;
+			const childTop = 100;
+
+			const iframe = document.createElement( 'iframe' );
+			parent.appendChild( iframe );
+			// JSDom doesn't have a layout engine
+			// so we need to mock getBoundingClientRect and DOMRect.
+			iframe.getBoundingClientRect = () => ( {
```
Comment (jeyip) [inline packages/dom/src/test/dom.js:230]: To make it clearer that we're mocking this function, and to fall more in line with idiomatic jest, could we use a jest method like `mockImplementation`? (not sure if that is the exact one)

```javascript
iframe.getBoundingClientRect = jest.mockImplementation(() => ( {
				width: 100,
				height: 100,
				top: iframeTop,
				left: iframeLeft,
			} ));
```

File: packages/dom/src/test/dom.js [hunk 4]
```diff
@@ -200,6 +201,70 @@ describe( 'DOM', () => {
...
+					right: left + width,
+					bottom: top + height,
+					width,
+					height,
+				};
+			};
+
+			const child = document.createElement( 'div' );
+			iframe.contentWindow.document.body.appendChild( child );
+			child.getBoundingClientRect = () => ( {
```
Comment (jeyip) [inline packages/dom/src/test/dom.js:254]: https://github.com/WordPress/gutenberg/pull/28630/files#r572190506 Same as above

**Review by jeyip — COMMENTED** [review]
+1 when discussions about `dom` method name and jest mock testing methods are resolved.

**Comment by ellatrix** [comment]
When did this regress? When I merged the iframe PR, it worked correctly. Would be great to know what changed broke it.

**Comment by david-szabo97** [comment]
> When did this regress? When I merged the iframe PR, it worked correctly. Would be great to know what changed broke it.

I don't think it was a regress. Popover works fine and offsets its position by the iframe position. But `InsertionPointPopover` used custom logic to calculate its position and I think the bug didn't surface when testing the iframe PR.

File: packages/dom/src/test/dom.js (hunk 3 above)
Comment (david-szabo97) [reply packages/dom/src/test/dom.js:230]: Great idea! Wrapped them in `jest.fn( impl )` which is basically a shorthand for `jest.fn().mockImplementation( impl )`.

File: packages/dom/src/dom.js (hunk 2 above)
Comment (david-szabo97) [reply packages/dom/src/dom.js:1015]: 🤔  How about: `offsetBoundsByOwnerFrame`, `offsetRectByOwnerFrame`, `frameToScreenPosition`, `frameToScreenBounds` ?

**Comment by youknowriad** [comment]
The custom logic to calculate position is new and I added it in #27860

File: packages/dom/src/dom.js (hunk 2 above)
Comment (ellatrix) [reply packages/dom/src/dom.js:1015]: Why does it need to be public? Can we not offset the iframe in the popover after we get the anchor rect? Why are we using `getAnchorRect` in the first place? That's supposed to be an old API to be deprecated at some point.

Comment (david-szabo97) [reply packages/dom/src/dom.js]: @ellatrix Moved `offsetIframe` back into the popover package. This time into the `utils`.

**Comment by david-szabo97** [comment]
Moved `offsetIframe` into `popover/utils` to avoid making this function public. Also moved tests so at least we now have test coverage.

**Review by jeyip — APPROVED** [review]
lgtm!

**Comment by vindl** [comment]
@youknowriad can we include this in 10.0? It looks pretty broken now.

**Comment by youknowriad** [comment]
yes, sounds good to me.

**Pull request merged**